libethereum/BlockChainSync.cpp

```diff
@@ -0,0 +1,781 @@
+	time_t now = time(0);
+	unsigned blockCount = c_chainReorgSize;
+	if (lastBlockTime > now)
+		clog(NetWarn) << "Clock skew? Latest block is in the future";
+	else
+		blockCount += (now - lastBlockTime) / (unsigned)c_durationLimit;
+	clog(NetAllDetail) << "Estimated hashes: " << blockCount;
+	return blockCount;
+}
+
+void BlockChainSync::requestBlocks(EthereumPeer* _peer)
+{
+	if (host().bq().knownFull())
+	{
+		clog(NetAllDetail) << "Waiting for block queue before downloading blocks";
+		m_lastActiveState = m_state;
+		pauseSync();
+		_peer->setIdle();
+		return;
+	}
+	_peer->requestBlocks();
+	if (_peer->m_asking != Asking::Blocks) //nothing to download
+	{
+		peerDoneBlocks(_peer);
+		if (downloadMan().isComplete())
+			completeSync();
+		return;
+	}
+}
+
+void BlockChainSync::onPeerBlocks(EthereumPeer* _peer, RLP const& _r)
+{
+	RecursiveGuard l(x_sync);
+	DEV_INVARIANT_CHECK;
+	unsigned itemCount = _r.itemCount();
+	clog(NetMessageSummary) << "Blocks (" << dec << itemCount << "entries)" << (itemCount ? "" : ": NoMoreBlocks");
+
+	_peer->setIdle();
+	if (m_state != SyncState::Blocks && m_state != SyncState::NewBlocks)
+		clog(NetWarn) << "Unexpected Blocks received!";
+	if (m_state == SyncState::Waiting)
+	{
+		clog(NetAllDetail) << "Ignored blocks while waiting";
+		return;
+	}
+
+	if (itemCount == 0)
+	{
+		// Got to this peer's latest block - just give up.
+		peerDoneBlocks(_peer);
+		if (downloadMan().isComplete())
+			completeSync();
+		return;
+	}
+
+	unsigned success = 0;
+	unsigned future = 0;
+	unsigned unknown = 0;
+	unsigned got = 0;
+	unsigned repeated = 0;
+	u256 maxUnknownNumber = 0;
+	h256 maxUnknown;
+
+	for (unsigned i = 0; i < itemCount; ++i)
+	{
+		auto h = BlockInfo::headerHash(_r[i].data());
+		if (_peer->m_sub.noteBlock(h))
+		{
+			_peer->addRating(10);
+			switch (host().bq().import(_r[i].data(), host().chain()))
+			{
+			case ImportResult::Success:
+				success++;
+				break;
+
+			case ImportResult::Malformed:
+			case ImportResult::BadChain:
+				_peer->disable("Malformed block received.");
+				return;
+
+			case ImportResult::FutureTimeKnown:
+				future++;
+				break;
+			case ImportResult::AlreadyInChain:
+			case ImportResult::AlreadyKnown:
+				got++;
+				break;
+
+			case ImportResult::FutureTimeUnknown:
+				future++; //Fall through
+
+			case ImportResult::UnknownParent:
+			{
+				unknown++;
+				if (m_state == SyncState::NewBlocks)
+				{
+					BlockInfo bi;
+					bi.populateFromHeader(_r[i][0]);
+					if (bi.number > maxUnknownNumber)
+					{
+						maxUnknownNumber = bi.number;
+						maxUnknown = h;
+					}
+				}
+				break;
+			}
+
+			default:;
+			}
+		}
+		else
+		{
+			_peer->addRating(0);	// -1?
+			repeated++;
+		}
+	}
+
+	clog(NetMessageSummary) << dec << success << "imported OK," << unknown << "with unknown parents," << future << "with future timestamps," << got << " already known," << repeated << " repeats received.";
+
+	if (host().bq().unknownFull())
```

Comment:
what's this about? it's very easy to have a message full of unknown blocks when syncing the chain in parallel across multiple peers.